shipyard_blueprints/microsoft-power-bi/shipyard_microsoft_power_bi/microsoft_power_bi_utils.py

```diff
@@ -0,0 +1,201 @@
+import msal
+import time
+
+from json import JSONDecodeError
+from shipyard_templates.exit_code_exception import ExitCodeException
+
+FAILED_JOB_STATUSES = {"Failed", "PartialComplete"}
+ONGOING_JOB_STATUSES = {"Queued", "InProgress", "Unknown"}
+COMPLETE_JOB_STATUSES = {"Completed", "Success"}
+
+
+def generate_access_token_from_client(bi_instance):
+    return msal.ConfidentialClientApplication(
+        client_id=bi_instance.client_id,
+        client_credential=bi_instance.client_secret,
+        authority=f"https://login.microsoftonline.com/{bi_instance.tenant_id}",
+    ).acquire_token_for_client(
+        scopes=["https://analysis.windows.net/powerbi/api/.default"]
+    )
+
+
+def generate_access_token_from_un_pw(bi_instance):
+    return msal.PublicClientApplication(
+        client_id=bi_instance.client_id,
+        authority=f"https://login.microsoftonline.com/{bi_instance.tenant_id}",
+    ).acquire_token_by_username_password(
+        username=bi_instance.username,
+        password=bi_instance.password,
+        scopes=["https://analysis.windows.net/powerbi/api/.default"],
+    )
+
+
+class PowerBiUtils:
```

Comment:
I'm not going to get hung up on it, but given that all the methods in here are static methods, it probably didn't need to be a class. Could have just be a set of functions